### PR TITLE
OneClick button cart page (#223)

### PR DESCRIPTION
**Pull request by rus-daniel**

### CODE REVIEW

File: app/design/frontend/base/default/template/oyst/oneclick/cart/oneclick.phtml [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+<div id="oyst-1click-button-wrapper">
+    <div id="oyst-1click-button"></div>
+</div>
+
+<script>
+    (function () {
+        if ("function" === typeof oystOneClick) {
+            oystOneClick(<?php echo $this->getOneClickUrl(); ?>);
+        } else {
+            console.warn("Missing function oystOneClick");
```
Comment (mehdichaouch) [inline app/design/frontend/base/default/template/oyst/oneclick/cart/oneclick.phtml]: change for `console.error(`

File: skin/frontend/base/default/js/oyst/cart/oneclick.js [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+        opts = opts || {};
+
+        ready(function () {
+            var form = new FormData();
+            form.append("preload", opts.preload);
+
+            var products = [];
+            var actions = document.getElementsByClassName("product-cart-actions");
+
+            for (var i = 0; i < actions.length; i++) {
```
Comment (mehdichaouch) [inline skin/frontend/base/default/js/oyst/cart/oneclick.js]: use `actions.forEach(...`

File: app/code/community/Oyst/OneClick/Model/System/Config/Source/CartPosition.php [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+    /**
+     * Button position from cart.
+     *
+     * @return array
+     */
+    public function toOptionArray()
+    {
+        return array(
+            array(
+                'value' => 1,
```
Comment (mehdichaouch) [inline app/code/community/Oyst/OneClick/Model/System/Config/Source/CartPosition.php]: Use explicit value

File: app/code/community/Oyst/OneClick/Model/System/Config/Source/CartPosition.php [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+{
+    /**
+     * Button position from cart.
+     *
+     * @return array
+     */
+    public function toOptionArray()
+    {
+        return array(
+            array(
```
Comment (mehdichaouch) [inline app/code/community/Oyst/OneClick/Model/System/Config/Source/CartPosition.php]: Add value for `Disable`

Default value is: Disabled

File: app/code/community/Oyst/OneClick/Model/System/Config/Source/CartPosition.php [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+                'value' => 1,
+                'label' => Mage::helper('oyst_oneclick')->__('Top'),
+            ),
+            array(
+                'value' => 0,
```
Comment (mehdichaouch) [inline app/code/community/Oyst/OneClick/Model/System/Config/Source/CartPosition.php]: Use explicit value

File: app/code/community/Oyst/OneClick/etc/system.xml [hunk 6]
```diff
@@ -593,6 +593,27 @@
                         </create>
                     </fields>
                 </new_customer_account>
+
+                <cart_button translate="label">
```
Comment (mehdichaouch) [inline app/code/community/Oyst/OneClick/etc/system.xml]: Change for `checkout_cart_button`

File: app/code/community/Oyst/OneClick/etc/system.xml [hunk 7]
```diff
@@ -593,6 +593,27 @@
...
+                    <label>Position of OneClick button from cart</label>
```
Comment (mehdichaouch) [inline app/code/community/Oyst/OneClick/etc/system.xml]: `Position of button from checkout cart`

File: app/code/community/Oyst/OneClick/etc/system.xml [hunk 8]
```diff
@@ -593,6 +593,27 @@
...
+                    <frontend_type>text</frontend_type>
+                    <sort_order>60</sort_order>
+                    <show_in_default>1</show_in_default>
+                    <show_in_website>1</show_in_website>
+                    <show_in_store>1</show_in_store>
+                    <fields>
+                        <position translate="label">
+                            <label>Select position of OneClick button</label>
```
Comment (mehdichaouch) [inline app/code/community/Oyst/OneClick/etc/system.xml]: `Select position of button`

File: app/code/community/Oyst/OneClick/etc/system.xml [hunk 9]
```diff
@@ -593,6 +593,27 @@
...
+                            <frontend_type>select</frontend_type>
+                            <source_model>oyst_oneclick/system_config_source_cartPosition</source_model>
+                            <config_path>oyst/cart_button/position</config_path>
```
Comment (mehdichaouch) [inline app/code/community/Oyst/OneClick/etc/system.xml]: `oyst/oneclick/checkout_cart_button_position`

File: app/design/frontend/base/default/layout/oyst_oneclick.xml [hunk 10]
```diff
@@ -41,4 +41,30 @@
...
+                <action method="setText" ifconfig="oyst/oneclick/enable">
+                    <text helper="oyst_oneclick/getOneClickJs"/>
+                </action>
+            </block>
+            <action method="addCss" ifconfig="oyst/oneclick/enable"><stylesheet>css/oyst/oneclick.css</stylesheet></action>
+            <action method="addItem"><type>skin_js</type><name>js/oyst/cart/oneclick.js</name></action>
+        </reference>
+
+        <reference name="checkout.cart.top_methods">
+            <action method="insert" ifconfig="oyst/cart_button/position">
```
Comment (mehdichaouch) [inline app/design/frontend/base/default/layout/oyst_oneclick.xml]: Change with new path

File: skin/frontend/base/default/js/oyst/cart/oneclick.js [hunk 11]
```diff
@@ -0,0 +1,138 @@
...
+            xhr.setRequestHeader("cache-control", settings.cacheControl);
+
+            xhr.onload = function () {
+                if (200 === xhr.status) {
+                    var data = JSON.parse(xhr.responseText);
+
+                    var isErrorsInForm = null;
+                    var messageProductViewElem = document.getElementById("messages_product_view");
+                    // If not the preload of button run form validation to know if they are errors
+                    if (!opts.preload) {
```
Comment (mehdichaouch) [inline skin/frontend/base/default/js/oyst/cart/oneclick.js]: Verify the CSS class and the necessity of the different control in this if block

File: skin/frontend/base/default/js/oyst/cart/oneclick.js (hunk 2 above)
Comment (rus-daniel) [reply skin/frontend/base/default/js/oyst/cart/oneclick.js]: Can't use forEach because it's an HTMLCollection

Comment (mehdichaouch) [reply skin/frontend/base/default/js/oyst/cart/oneclick.js]: Ok ;)

File: app/locale/fr_FR/Oyst_OneClick.csv [hunk 12]
```diff
@@ -151,3 +151,9 @@
 
 "Unable to refund Oyst order %s (%s - %s).","Impossible de rembourser de la commande %s (%s - %s)."
 "There is already one credit memo in the queue.","Il y a déjà un remboursement en cours."
+
+"Top","Haut"
+"Bottom","Bas"
+"Disabled","Désactivée"
+"Position of button from checkout cart","Position du bouton de la caisse panier"
```
Comment (mehdichaouch) [inline app/locale/fr_FR/Oyst_OneClick.csv]: Position du bouton dans la page panier

File: app/code/community/Oyst/OneClick/etc/system.xml (hunk 6 above)
Comment (mehdichaouch) [inline app/code/community/Oyst/OneClick/etc/system.xml]: Change for `checkout_cart_button`

Comment (mehdichaouch) [reply app/code/community/Oyst/OneClick/etc/system.xml]: @rus-daniel could you change this

File: skin/frontend/base/default/js/oyst/cart/oneclick.js [hunk 13]
```diff
@@ -0,0 +1,88 @@
...
+            xhr.open(settings.method, settings.url, settings.async);
+            xhr.setRequestHeader("cache-control", settings.cacheControl);
+
+            xhr.onload = function () {
+                if (200 === xhr.status) {
+                    var data = JSON.parse(xhr.responseText);
+
+                    var isErrorsInForm = null;
+                    var messageProductViewElem = document.getElementById("messages_product_view");
+                    // If not the preload of button run form validation to know if they are errors
```
Comment (mehdichaouch) [inline skin/frontend/base/default/js/oyst/cart/oneclick.js]: Useless comment

File: skin/frontend/base/default/js/oyst/cart/oneclick.js [hunk 14]
```diff
@@ -0,0 +1,88 @@
...
+            var xhr = new XMLHttpRequest();
+            xhr.open(settings.method, settings.url, settings.async);
+            xhr.setRequestHeader("cache-control", settings.cacheControl);
+
+            xhr.onload = function () {
+                if (200 === xhr.status) {
+                    var data = JSON.parse(xhr.responseText);
+
+                    var isErrorsInForm = null;
+                    var messageProductViewElem = document.getElementById("messages_product_view");
```
Comment (mehdichaouch) [inline skin/frontend/base/default/js/oyst/cart/oneclick.js]: Useless

File: skin/frontend/base/default/js/oyst/cart/oneclick.js [hunk 15]
```diff
@@ -0,0 +1,88 @@
...
+                    // If not the preload of button run form validation to know if they are errors
+
+                    cb(isErrorsInForm, data.url);
+                    if (messageProductViewElem) {
```
Comment (mehdichaouch) [inline skin/frontend/base/default/js/oyst/cart/oneclick.js]: Useless if block

File: skin/frontend/base/default/js/oyst/cart/oneclick.js [hunk 16]
```diff
@@ -0,0 +1,88 @@
...
+                data: form,
+                method: "POST",
+                mimeType: "multipart/form-data",
+                url: oneClickUrl
+            };
+
+            var xhr = new XMLHttpRequest();
+            xhr.open(settings.method, settings.url, settings.async);
+            xhr.setRequestHeader("cache-control", settings.cacheControl);
+
```
Comment (mehdichaouch) [inline skin/frontend/base/default/js/oyst/cart/oneclick.js]: remove this breakline

File: app/design/frontend/base/default/layout/oyst_oneclick.xml [hunk 17]
```diff
@@ -41,4 +41,34 @@
...
+                <action method="setText" ifconfig="oyst/oneclick/enable">
+                    <text helper="oyst_oneclick/getOneClickJs"/>
+                </action>
+            </block>
+            <action method="addCss" ifconfig="oyst/oneclick/enable"><stylesheet>css/oyst/oneclick.css</stylesheet></action>
+            <action method="addItem"><type>skin_js</type><name>js/oyst/cart/oneclick.js</name></action>
+        </reference>
+    </checkout_cart_index>
+
+    <oneclick_cart_top>
```
Comment (mehdichaouch) [inline app/design/frontend/base/default/layout/oyst_oneclick.xml]: rename with `oneclick_checkout_cart_top`

File: app/design/frontend/base/default/layout/oyst_oneclick.xml [hunk 18]
```diff
@@ -41,4 +41,34 @@
...
+        <reference name="checkout.cart.top_methods">
+            <action method="insert">
+                <block>oyst.oneclick.cart.checkout</block>
+            </action>
+        </reference>
+    </oneclick_cart_top>
+
+    <oneclick_cart_bottom>
```
Comment (mehdichaouch) [inline app/design/frontend/base/default/layout/oyst_oneclick.xml]: Rename with `oneclick_checkout_cart_bottom`

File: app/code/community/Oyst/OneClick/Model/Observer.php [hunk 19]
```diff
@@ -111,4 +111,21 @@ public function redirectToNotificationsGrid(Varien_Event_Observer $observer)
...
+
+    public function appendOneClickButton(Varien_Event_Observer $observer)
+    {
+        if ('checkout_cart_index' === $observer->getAction()->getFullActionName()) {
+            $layout = $observer->getLayout();
+            $position = Mage::getStoreConfig('oyst/oneclick/checkout_cart_button_position');
+
+            switch ($position) {
+                case 'top':
+                    $layout->getUpdate()->addHandle('oneclick_cart_top');
```
Comment (mehdichaouch) [inline app/code/community/Oyst/OneClick/Model/Observer.php]: Rename with `oneclick_checkout_cart_bottom`

File: app/code/community/Oyst/OneClick/Model/Observer.php [hunk 20]
```diff
@@ -111,4 +111,21 @@ public function redirectToNotificationsGrid(Varien_Event_Observer $observer)
...
+                    break;
+                case 'bottom':
+                    $layout->getUpdate()->addHandle('oneclick_cart_bottom');
```
Comment (mehdichaouch) [inline app/code/community/Oyst/OneClick/Model/Observer.php]: Rename with `oneclick_checkout_cart_bottom `

File: app/code/community/Oyst/OneClick/Model/Observer.php [hunk 21]
```diff
@@ -111,4 +111,21 @@ public function redirectToNotificationsGrid(Varien_Event_Observer $observer)
                 ->sendHeadersAndExit();
         }
     }
+
+    public function appendOneClickButton(Varien_Event_Observer $observer)
```
Comment (mehdichaouch) [inline app/code/community/Oyst/OneClick/Model/Observer.php]: rename with `appendCheckoutCartOneClickButton`

**Comment by mehdichaouch** [comment]
@rus-daniel could you rebase it from master. Thank you.

File: app/code/community/Oyst/OneClick/etc/system.xml [hunk 22]
```diff
@@ -604,18 +604,39 @@
...
-                            <frontend_type>text</frontend_type>
-                            <config_path>oyst/oneclick/order_delay</config_path>
+                        <label>Delay before the order is placed</label>
+                        <comment>In minutes</comment>
+                        <frontend_type>text</frontend_type>
+                        <config_path>oyst/oneclick/order_delay</config_path>
+                        <sort_order>1</sort_order>
+                        <show_in_default>1</show_in_default>
+                        <show_in_website>1</show_in_website>
+                        <show_in_store>1</show_in_store>
```
Comment (mehdichaouch) [inline app/code/community/Oyst/OneClick/etc/system.xml]: @rus-daniel re-indent lines above on `<time>`

File: app/code/community/Oyst/OneClick/etc/system.xml [hunk 23]
```diff
@@ -616,6 +616,27 @@
                         </time>
                     </fields>
                 </order_delay>
+
+                <checkout_cart_button translate="label">
```
Comment (mehdichaouch) [inline app/code/community/Oyst/OneClick/etc/system.xml]: Change for `customization_button_checkout_cart` like that all new customization fields will move here.

**Pull request merged**